apple/login.go

```diff
@@ -0,0 +1,285 @@
+package apple
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"time"
+
+	"github.com/dghubble/gologin/v2"
+	oauth2Login "github.com/dghubble/gologin/v2/oauth2"
+
+	"github.com/dgrijalva/jwt-go"
+
+	"golang.org/x/oauth2"
+)
+
+// Apple base URL
+const (
+	AppleBaseURL = "https://appleid.apple.com"
+)
+
+// User is an Apple user.
+//
+// Note that user IDs are unique to each app.
+type User struct {
```

Comment:
Even though not explicitly stated in their docs (that I could find), what Apple sends back is an OpenID Connect token -- I am guessing that some of the other providers you support do the same, but that it is effectively ignored in favour of verification using the access token and calling a userinfo-type API.

I think it's safe to assume that the OIDC token returned by Apple is standard, as it would be by some of the other providers. Perhaps instead of using jwt-go to parse and validate it, I could look into using something specifically designed for standard OIDC parsing and validation purposes? (e.g., https://github.com/coreos/go-oidc/ ?)

Unfortunately I can't control what Apple provides in terms of libraries, so I think the real question is whether you think gologin should be responsible for understanding, validating and parsing user information from an OIDC token supplied by an auth provider, or whether it should punt that to the developer.  If you think it should punt, then I'm curious as to why you don't also punt to the developer for the other providers for e.g., using the access token to call /me (in the case of Facebook), /userinfo (in the case of Google), etc.

Reply:
Chained handlers let each perform a step in the process. Handlers for current providers add some minimal user representation (not always _the_ provider's hydrated representation) to the context. Its done because those providers maintain/version libraries for that lookup/exchange or the exchange is very simple, which makes it reasonable enough to provide the convenience. I'd like Apple's token to user exchange logic to reside in a library Apple maintains over time. 

For now, you can limit the login handler to just adding the token to the context. Leave the token to user exchange to a separate handler (users bring their own). Then you (or others) can iterate on that handler (DIY JWT vs OIDC). You (or others) could have a package. Or maybe in future one of those is brought into mainline.

Main hope would be to get this PR down to something minimal. If the context gets someone part way and the last handler is a exercise for the user / link to an example, that's ok.

Reply:
The internal oauth1 and oauth2 handlers are exposed so another reasonable option is actually that you publish / maintain Apple handlers too. And import. You can take whatever approaches that way and it sounds like you're more likely to use/test Apple login.